### DISCUSSION TITLE
The client_exited life cycle event seems to be called for all class instances

### DISCUSSION THREAD

**Discussion by MattTheCuber** [Kitware/trame#475] [General] [answered]
We have a class that opens on a new route that the main application calls using `template_name`. We wanted to attach a listener for when the client disconnects from one of these viewers to cleanup. However, when we initialize multiple of these and connect the client to each, then disconnect from one of the views, the `client_exited` event is called for each of our class instances.

Snippet:
```python
@TrameApp()
class Viewer:
    def __init__(self, template_name):
        server = get_server(client_type="vue3")
        with VAppLayout(server, template_name=template_name):
            ...

    @life_cycle.client_exited
    def cleanup(self):
        print(id(self)) # This prints multiple different ids
```

Is there a way to know which window was closed by the client or to only attach the listener to the current class instance?

**Comment by jourdain** [comment]
Those life_cycle events are bound to a server instance, hence the issue. 
But if you want to be individually aware when a given client disconnects from its personal layout, you can do that via a [LifeCycle component](https://github.com/Kitware/trame-components/blob/master/trame_components/widgets/trame.py#L97) and trigger a method on the server side with the client/template name.

**Reply by MattTheCuber** [reply]
@jourdain We dug into this and found that the `destroyed` and `before_destroy` events are not triggered when the client closes their connection. We actually have been unable to trigger any events outside of `created` and `mounted`. Do you have any insight on this issue?

We also found `ClientTriggers` provides similar functionality to the `LifeCycleMonitor` component but interfaces with Python better (see [this example](https://github.com/Kitware/trame/blob/b66427334d9fda8e272c1c454387c15a9c2a3f4a/examples/validation/core/02_trigger_client.py#L50)). The events behaved the same way as the `LifeCycleMonitor`. However, I think that example has a bug as it uses `beforeDestroy` instead of `before_destroy`.

**Reply by jourdain** [reply]
The ClientTriggers give you indeed more options than just the Vue component cycle. In other words it kind superseed it if you just want the create/mounted/delete. 

I recently moved those helper components to trame-client and I've [tested them](https://github.com/Kitware/trame-client/blob/master/examples/vue3/life_cycle_client.py).  And I see the destroy when the template rebuild. Maybe your issue is that you are closing the window/app rather than nicely unmounting the vue component.

**Reply by MattTheCuber** [reply]
That was the goal for this issue. We were hoping to receive an event when the client closes the window/app so that we can cleanup initialized classes and data in RAM. Is that not what the `ClientTriggers` and `LifeCycleMonitor` are used for? I guess a better question is: When are each of the events called?

**Reply by jourdain** [reply]
Those are tight to the Vue component level. The delete will be trigger when you switch route in the router or when the template redraw. The mount event will signal that a client connected and the UI is up. For the close of the window, we just have the global `ctrl.on_client_exited`. But you won't know which client if you have more than one.

What you could do is keep a counter around and do a +1 on `on_client_connected` and -1 on `on_client_exited` and when you reach 0, you do your garbage collection. But keep in mind, the process will automatically exit once a timeout after all the connections are gone.

**Reply by MattTheCuber** [reply]
That is unfortunate. I don't think the `on_client_exited` would work very well as the `on_client_exited` gets called for each open viewer, not once. So this would decrement N number of times resulting in a value less than 0.

**Reply by jourdain** [reply]
Then you need to put your own method call on `beforeunload`. 

I think I finally get it. As you are reusing the same server on which you bind several instance of apps. And you want a specific cleanup method to be called based on which template got closed...

Ideally if you could run the following JS function on `beforeunload` you might be able to get the destroy event.

```js
trame.set(templateName, "<div>Exiting</div>")
```

I will need to investigate and spend some time to properly provide a good solution for your per template, cleanup method.

**Reply by jourdain** [reply]
Another option would be to extend https://github.com/Kitware/trame-client/blob/master/vue3-app/src/components/TrameApp.js#L108-L110 to allow other network calls which could be unique per template...

**Reply by jourdain** [reply]
Keep thinking out loud here... We could add `beforeunload` on `ClientTriggers`.

**Reply by MattTheCuber** [reply]
> I think I finally get it. As you are reusing the same server on which you bind several instance of apps. And you want a specific cleanup method to be called based on which template got closed...

Should we be using a new server for each viewer that the client requests?

**Reply by MattTheCuber** [reply]
Adding `beforeunload` to `ClientTriggers` would be great.

**Reply by jourdain** [reply, marked as answer]
`trame-client>=3.0.1` added a __exit=__ event on which you can bind a method that will be tight to your UI instance. 
See [this example for reference](https://github.com/Kitware/trame-client/blob/master/examples/vue3/trigger.py#L63).

**Reply by MattTheCuber** [reply]
Awesome, thanks Sebastien! I will try this out Monday.

**Reply by MattTheCuber** [reply]
This worked great, thanks!
